File: Tp02/TP02Q09 - Heapsort em Java/Principal.java
import java.util.*;

class Jogador{

    private int id,
                altura,
                peso,
                anoNascimento;
    private String nome,
                   universidade,
                   cidadeNascimento,
                   estadoNascimento;

    public Jogador(){

    }
    public Jogador(int id, String nome, int altura,int peso,String universidade,int anoNascimento,String cidadeNascimento,String estadoNascimento){

        setId(id);
        setNome(nome);
        setAltura(altura);
        setPeso(peso);
        setUniversidade(universidade);
        setAnoNascimento(anoNascimento);
        setCidadeNascimento(cidadeNascimento);
        setEstadoNascimento(estadoNascimento);
    }
    public int getId(){
        return id;

    }
    public void setId(int id){
        if(id >= 0)
            this.id = id;
    }
    public String getNome(){
        return nome;

    }
    public void setNome(String nome){
        nome = validaDado(nome);

        this.nome = nome;
    }
    public int getAltura(){

        return altura;

    }
    public void setAltura(int altura){

        this.altura = altura;
    }
    public int getPeso(){

        return peso;

    }
    public void setPeso(int peso){

        this.peso = peso;


    }
    public String getUniversidade(){
        return universidade;
        
    }
    public void setUniversidade(String universidade){
        universidade = validaDado(universidade);
        this.universidade = universidade;
    }
    public int getAnoNascimento(){
        return anoNascimento;

    }
    public void setAnoNascimento(int anoNascimento){
        this.anoNascimento = anoNascimento;

    }
    public String getCidadeNascimento(){
        return cidadeNascimento;
        
    }
    public void setCidadeNascimento(String cidadeNascimento){
        cidadeNascimento = validaDado(cidadeNascimento);
        this.cidadeNascimento = cidadeNascimento;    

    }
    public String getEstadoNascimento(){
        return estadoNascimento;
        
    }
    public void setEstadoNascimento(String estadoNascimento){
        estadoNascimento = validaDado(estadoNascimento);
    
        this.estadoNascimento = estadoNascimento;
    }
    public String validaDado(String dado){

        if(dado.isEmpty())
            dado = "nao informado";

        return dado;

    }
    /**
     *Imprime a instancia
     * */

    public void imprimir(){

        MyIO.println("[" + getId() + " ## " + getNome() + " ## " +getAltura() + " ## " + getPeso() + " ## " + getAnoNascimento() + " ## " + getUniversidade() + " ## " + getCidadeNascimento() + " ## "+ getEstadoNascimento() +"]");
    }

    public void ler(String jogador){

        String[] valores = jogador.split(",",8);

        setId(valores[0]!="" ?Integer.parseInt(valores[0]): -1);
        setNome(valores[1]);//.replaceAll("\\*",""));
        setAltura(Integer.parseInt(valores[2]));
        setPeso(Integer.parseInt(valores[3]));
        setUniversidade(valores[4]);
        setAnoNascimento(Integer.parseInt(valores[5]));
        setCidadeNascimento(valores[6]);
        setEstadoNascimento(valores[7]);

    }
    public void  clone(Jogador resultado){
        resultado.setId(getId());
        resultado.setNome(getNome());
        resultado.setAltura(getAltura());
        resultado.setPeso(getPeso());
        resultado.setUniversidade(getUniversidade());
        resultado.setAnoNascimento(getAnoNascimento());
        resultado.setCidadeNascimento(getCidadeNascimento());
        resultado.setEstadoNascimento(getEstadoNascimento());

        

    }      
    
    /**
    Classe Principal: Coordena as informações e a classe acima
    
    **/  

    }
    public class Principal{
	   protected static Jogador[] jogadores;
	   protected static Jogador[] novoVetor;
	   protected static int comp;

        /**
         *@param vetor vazio

         metodo le o arquivo e preenche o vetor
         * */
        public static void leArquivoInicial(Jogador[] jogadores){

            Arq.openRead("/tmp/players.csv");
            String dados = "";
            int i = 0;

            Arq.readLine();//remove linha de cabecalho
            


            while(i<3922){
                dados = Arq.readLine();
                jogadores[i] = new Jogador();
                jogadores[i].ler(dados);

                i++;
            }
        }
        
        public static void criaLog(long tempo){


			Arq.openWrite("698159_heapsort.txt");
			Arq.println("698159\t" + tempo +"ms" + "\t" + comp);
			Arq.close();


		}
        
        public static void SelecionaNovoVetor(int id, Jogador Novo){
        
        	jogadores[id].clone(Novo);
        
        
        }
        public static boolean isFim(String palavra){

            return palavra.length() == 3 && palavra.charAt(0)=='F' && palavra.charAt(1)=='I' && palavra.charAt(2)=='M';
        }
        
        /**
        ordena subArray
        @param numero de posicoes no array
        **/
        
        public static void heapsort(int n) {
      		//Alterar o vetor ignorando a posicao zero
      		Jogador[] tmp = new Jogador[n+1];
      		for(int i = 0; i < n; i++){
      			tmp[i+1] = new Jogador();
        		 novoVetor[i].clone(tmp[i+1]);
     		 }
     		 novoVetor = tmp;

     		 //Contrucao do heap
      		for(int tamHeap = 2; tamHeap <= n; tamHeap++){
       		  construir(tamHeap);
    	    }

      		//Ordenacao propriamente dita
     		int tamHeap = n;
      		while(tamHeap > 1){
         		swapJogador(1, tamHeap--);
         		reconstruir(tamHeap);
      		}

		    //Alterar o vetor para voltar a posicao zero
      		tmp = novoVetor;
	      	novoVetor = new Jogador[n];
      		for(int i = 0; i < n; i++){
      			novoVetor[i] = new Jogador();
         		tmp[i+1].clone(novoVetor[i]);
      		}
   		}

		public static void construir(int tamHeap){
      		for(int i = tamHeap; i > 1 && comparaJogadores(i) ; i /= 2){
         	swapJogador(i, i/2);
      		}
   		}


   		public static void reconstruir(int tamHeap){
      		int i = 1;
      		while(i <= (tamHeap/2)){
         		int filho = getMaiorFilho(i, tamHeap);
         		if(novoVetor[i].getAltura() < novoVetor[filho].getAltura() ||(novoVetor[i].getAltura() == novoVetor[filho].getAltura() && novoVetor[i].getNome().compareTo(novoVetor[filho].getNome()) <0)){
            		swapJogador(i, filho);
            		i = filho;
         		}else{
            		i = tamHeap;
         		}
      		}
   		}

   		public static int getMaiorFilho(int i, int tamHeap){
      		int filho;
      		if ((2*i == tamHeap || novoVetor[2*i].getAltura() > novoVetor[2*i+1].getAltura()) || (novoVetor[2*i].getAltura() == novoVetor[2*i+1].getAltura() && novoVetor[2*i].getNome().compareTo(novoVetor[2*i+1].getNome()) > 0)){
         		filho = 2*i;
      		} else {
         		filho = 2*i + 1;
      		}
      		return filho;
   		}
   		
   		
   		/**
   			swap modificado para instancias de jogadores
   		**/
   		public static void swapJogador(int pos1,int pos2){
   		
   			Jogador tmp = new Jogador();
   			novoVetor[pos1].clone(tmp);
   			novoVetor[pos2].clone(novoVetor[pos1]);
   			tmp.clone(novoVetor[pos2]);
   			
   		
   		}
   		
   		public static boolean comparaJogadores(int i){
        
        	
        	comp++;
        	return ((novoVetor[i].getAltura()>novoVetor[i/2].getAltura())||(novoVetor[i].getAltura()==novoVetor[i/2].getAltura()&&novoVetor[i].getNome().compareTo(novoVetor[i/2].getNome())>0));
        
        }
   		
   		
   		public static long now(){


			return new Date().getTime();

		}

        


        public static void main(String[] Args){
            int tam = 3922;

            jogadores = new Jogador[tam];//vetor original
            novoVetor = new Jogador[tam];//vetor subsequente
            
            comp = 0;//numero de comparacoes


            leArquivoInicial(jogadores);
            String palavra = MyIO.readString();
            int posicao = 0;
            int n = 0;

            while(!isFim(palavra)){
                posicao = Integer.parseInt(palavra);
                
                novoVetor[n] = new Jogador();
                SelecionaNovoVetor(posicao,novoVetor[n++]);
                
                palavra = MyIO.readString();
            }
            long inicio = now();
            heapsort(n);
			long fim = now();
			long tempo = (fim - inicio);
			criaLog(tempo);
			for(int i = 0;i<n;i++){
			
				novoVetor[i].imprimir();
			
			}



        }






    }





















    

File: Tp02/TP02Q17 - Heapsort PARCIAL em C/principal.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

typedef struct
{

    int id,
        altura,
        peso,
        anoNascimento;

    char *nome,
        *universidade,
        *cidadeNascimento,
        *estadoNascimento;

} jogador;

typedef jogador *PJogador;
typedef jogador NJogador;

void construtor1(PJogador jogador)
{

    jogador->id = 0;
    jogador->altura = 0;
    jogador->peso = 0;
    jogador->anoNascimento = 0;

    jogador->nome = (char *)malloc(sizeof(char) * 100);
    jogador->universidade = (char *)malloc(sizeof(char) * 100);
    jogador->cidadeNascimento = (char *)malloc(sizeof(char) * 100);
    jogador->estadoNascimento = (char *)malloc(sizeof(char) * 100);
}

/**
	retira possiveis chiados nas strings
**/
char *trataAsterisco(char *str)
{

    if (strstr(str, "*") != NULL)
    {

        for (int i = 0; i < strlen(str); i++)
        {
            if (str[i] == '*')
            {
                str[i] = '\0';
            }
        }
    }

    return str;
}
/**
	divide a entrada inteira em partes menores e as le para o jogador
**/

void ler(PJogador jogador, char dados[])
{

    char *array[8];

    char *emBranco = (char *)malloc(sizeof(char *) * 100);
    strcpy(emBranco, "nao informado");

    array[0] = strsep(&dados, ",");

    for (int i = 1; i < 8; i++)
    {
        array[i] = strsep(&dados, ",");
    }

    //array[1] = trataAsterisco(array[1]);

    jogador->id = atoi(array[0]);
    jogador->altura = atoi(array[2]);
    jogador->peso = atoi(array[3]);
    jogador->anoNascimento = atoi(array[5]);

    if (strcmp(array[1], "\0") == 0)
        jogador->nome = emBranco;
    else
        jogador->nome = array[1];

    if (strcmp(array[4], "\0") == 0)
        jogador->universidade = emBranco;
    else
        jogador->universidade = array[4];

    if (strcmp(array[6], "\0") == 0)
    {
        jogador->cidadeNascimento = emBranco;
    }
    else
    {
        jogador->cidadeNascimento = array[6];
    }

    if (strcmp(array[7], "\n") == 0)
        jogador->estadoNascimento = emBranco;
    else
        jogador->estadoNascimento = strsep(&array[7], "\n");
}

/**
	Imprime instancia de Jogador selecionado nos moldes escolhidos
**/

void imprimir(PJogador jogador)
{

    printf("[%d ## %s ## %d ## %d ## %d ## %s ## %s ## %s]\n", jogador->id, jogador->nome, jogador->altura, jogador->peso, jogador->anoNascimento, jogador->universidade, jogador->cidadeNascimento, jogador->estadoNascimento);
}

void construtor2(PJogador jogador, char dados[])
{
    jogador->id = 0;
    jogador->altura = 0;
    jogador->peso = 0;
    jogador->anoNascimento = 0;

    jogador->nome = (char *)malloc(sizeof(char) * 100);
    jogador->universidade = (char *)malloc(sizeof(char) * 100);
    jogador->cidadeNascimento = (char *)malloc(sizeof(char) * 100);
    jogador->estadoNascimento = (char *)malloc(sizeof(char) * 100);
    ler(jogador, dados);
}

/**
	Clona instancia passada para nova posicao
**/

void clone(PJogador origem, PJogador resultado)
{
    construtor1(resultado);
    resultado->id = origem->id;
    strcpy(resultado->nome, origem->nome);
    resultado->altura = origem->altura;
    resultado->peso = origem->peso;
    strcpy(resultado->universidade, origem->universidade);
    resultado->anoNascimento = origem->anoNascimento;
    strcpy(resultado->cidadeNascimento, origem->cidadeNascimento);
    strcpy(resultado->estadoNascimento, origem->estadoNascimento);
}

/**

	le todos os integrantes do arquivo e os instancia como jogador
**/

void leArquivoInicial(PJogador jogador[])
{
    FILE *Arq = fopen("/tmp/players.csv", "r");

    char *dados = (char *)malloc(sizeof(char) * 200);
    int i = 0;
    fgets(dados, 200, Arq);

    while (i < 3922)
    {
        dados = (char *)malloc(sizeof(char) * 200);
        fgets(dados, 200, Arq);

        jogador[i] = (PJogador)malloc(sizeof(NJogador));
        construtor2(jogador[i], dados);

        i++;
    }
    fclose(Arq);
}

bool isFim(char *str)
{

    return strlen(str) == 3 && str[0] == 'F' && str[1] == 'I' && str[2] == 'M';
}
/*
	cria um array secundario com os jogadores indicados
*/
void criaSubArray(PJogador jogadores[], PJogador subArray[], int *n)
{
    char entrada[100];
    scanf(" %[^\n]", entrada);
    int pos;

    while (!isFim(entrada))
    {

        pos = atoi(entrada);
        subArray[*n] = (PJogador)malloc(sizeof(NJogador));
        clone(jogadores[pos], subArray[*n]);

        scanf(" %[^\n]", entrada);
        *n = *n + 1;
    }
}

void swapJogador(PJogador jogador1, PJogador jogador2)
{

    PJogador tmp = (PJogador)malloc(sizeof(NJogador));
    //printf("----------------------Antes-------------------\n");
    clone(jogador1, tmp);
    clone(jogador2, jogador1);
    clone(tmp, jogador2);
    //printf("----------------------Depois-------------------\n");

    free(tmp);
}

void construir(PJogador *array, int tamHeap,int *comp)
{
    for (int i = tamHeap; i > 1 && ((array[i]->altura < array[i / 2]->altura) || (array[i]->altura == array[i / 2]->altura && strcmp(array[i]->nome, array[i / 2]->nome) < 0)); i /= 2,*comp=*comp+1)
    {
    	
        swapJogador(array[i], array[i / 2]);
    }
}
//=============================================================================ok
int getMenorFilho(PJogador *array, int i, int tamHeap,int *comp)
{
    int filho;
    *comp=*comp+1;
    if (2 * i + 1 == tamHeap || ((array[2 * i]->altura > array[2 * i + 1]->altura) || (array[2 * i]->altura == array[2 * i + 1]->altura && strcmp(array[2 * i]->nome, array[2 * i + 1]->nome) > 0)))
    {
        filho = 2 * i + 1;
    }
    else
    {
        filho = 2 * i;
    }
    return filho;
}
//=============================================================================ok
void reconstruir(PJogador *array, int tamHeap,int *comp)
{
    int i = 1;
    while (i <= (tamHeap / 2))
    {
    	*comp=*comp+1;
        int filho = getMenorFilho(array, i, tamHeap,comp);
        if ((array[i]->altura > array[filho]->altura) || (array[i]->altura == array[filho]->altura && strcmp(array[i]->nome,array[filho]->nome)>0))
        {
            swapJogador(array[i], array[filho]);
            i = filho;
        }
        else
        {
            i = tamHeap;
        }
    }
}
//=============================================================================ok
void heapsortParcial(PJogador *array, int n, int k, int *comp)
{
    //Alterar o vetor ignorando a posicao zero
    PJogador arrayTmp[n + 1];
    for (int i = 0; i < n; i++)
    {
        arrayTmp[i + 1] = (PJogador)malloc(sizeof(NJogador));
        clone(array[i], arrayTmp[i + 1]);
    }

    //Contrucao do heap
    for (int tamHeap = 2; tamHeap <= n; tamHeap++)
    {
        construir(arrayTmp, tamHeap,comp);
    }
    int dir;
    int aux = 0;
    //Ordenacao propriamente dita
    int tamHeap = n;
    while (aux < k)
    {
        clone(arrayTmp[1], array[aux]);
        swapJogador(arrayTmp[1], arrayTmp[tamHeap--]);
        reconstruir(arrayTmp, tamHeap,comp);
        aux++;
    }
}

/**
   			swap modificado para instancias de jogadores
   		**/

void criaLog(double tempo, int comp)
{

    FILE *fp = fopen("698159_heapsortParcial.txt", "w");
    fprintf(fp, "698159 \t %.4fS \t %i ", tempo, comp);

    fclose(fp);
}

int main()
{

    PJogador jogadores[3922]; //array principal
    PJogador subArray[3922];  //array criado secundariamente

    int n = 0; //numero de posicoes do array secundario
    int comp = 0;

    clock_t inicio, fim; //calcular tempo

    int pos = 0;
    char *resp = (char *)malloc(sizeof(char) * 3);
    char *entrada = (char *)malloc(sizeof(char) * 100);

    leArquivoInicial(jogadores);

    criaSubArray(jogadores, subArray, &n);

    inicio = clock();
    int k = 10;
    heapsortParcial(subArray, n, k, &comp);
    fim = clock();
    for (int i = 0; i < k; i++)
    {

        imprimir(subArray[i]);
    }
    double total = (fim - inicio) / (double)CLOCKS_PER_SEC;
    criaLog(total, comp);

    return 0;
}
